src/workerd/api/pyodide/pyodide.c++

```diff
@@ -56,6 +58,32 @@ int PyodideMetadataReader::read(jsg::Lock& js, int index, int offset, kj::Array<
   return toCopy;
 }
 
+int PyodideMetadataReader::readMemorySnapshot(int offset, kj::Array<kj::byte> buf) {
+  int snapshotSize = memorySnapshot.size();
+  if (offset >= snapshotSize || offset < 0) {
+    return 0;
+  }
+  int toCopy = buf.size();
+  if (snapshotSize - offset < toCopy) {
+    toCopy = snapshotSize - offset;
+  }
+  memcpy(buf.begin(), &memorySnapshot[0] + offset, toCopy);
+  return toCopy;
+}
+
+int ArtifactBundler::readMemorySnapshot(int offset, kj::Array<kj::byte> buf) {
+  int snapshotSize = existingSnapshot.size();
+  if (offset >= snapshotSize || offset < 0) {
+    return 0;
+  }
+  int toCopy = buf.size();
+  if (snapshotSize - offset < toCopy) {
+    toCopy = snapshotSize - offset;
+  }
+  memcpy(buf.begin(), &existingSnapshot[0] + offset, toCopy);
+  return toCopy;
+}
```

Comment:
These two are practically identical, can we refactor to avoid duplication fairly easily here?